File: 프로그래머스/코딩테스트_고득점_kit/스택_큐/프린터/dailyco.js
/*
작성자 : dailyco
문제 난이도: 하
해결시간: 00:30:00
문제 링크 : https://programmers.co.kr/learn/courses/30/lessons/42587
블로그 링크 : https://dailyco.github.io/posts/printer/
*/

function solution(priorities, location) {
  var answer = 0;

  while (priorities.length) {
    var J = priorities.shift();
    var isPrint = true;
    for (let i = 0; i < priorities.length; i++) {
      if (J < priorities[i]) {
        isPrint = false;
        break;
      }
    }

    if (isPrint) {
      answer++;
      if (location === 0) return answer;
      else location--;
    } else {
      if (location === 0) location = priorities.length;
      else location--;
      priorities.push(J);
    }
  }

  return answer;
}

File: 프로그래머스/코딩테스트_고득점_kit/스택_큐/프린터/dailyco.cpp
/*
작성자 : dailyco
문제 난이도: 하
해결시간: 02:00:00
문제 링크 : https://programmers.co.kr/learn/courses/30/lessons/42587
블로그 링크 : https://dailyco.github.io/posts/printer/
*/

#include <vector>
#include <algorithm>
#include <list>

using namespace std;

int solution(vector<int> priorities, int location)
{
    list<int> ps(priorities.begin(), priorities.end());
    int p = priorities[location];
    int max;
    int answer = 0;

    max = *max_element(ps.begin(), ps.end());
    while (true)
    {
        int front = ps.front();
        if (max != front)
        {
            ps.pop_front();
            ps.push_back(front);
        }
        else
        {
            ps.pop_front();
            answer++;

            if (location == 0)
                break;
            max = *max_element(ps.begin(), ps.end());
        }

        if (--location < 0)
            location = ps.size() - 1;
    }

    return answer;
}